### PR TITLE
Google Analytics driver

### PR DESCRIPTION
**Pull request by tlrobinson**
#3020
- ~~Improve setup flow, ideally user:~~
  - ~~won't have to copy OAuth code~~
  - ~~can select which org they want from a dropdown~~
- [x] Figure out how to organize tables/fields/etc. There are a ton of metrics and dimensions. Currently everything is in one "table".
  - [x] should we show everything or only certain dimensions / metrics? (many aren't relevant if Google advertising products aren't used, for example)
    - currently showing everything except these https://github.com/metabase/metabase/pull/3619/commits/5ef9fb909a07ded86186723f4179a96fb010ac07
  - ~~some other way of organizing them?  Could we make pseudo-tables like User, Session, etc and use that to organize them?~~
- [x] Figure out how we want aggregations/metrics to work. GA has a [ton of metrics built in](https://developers.google.com/analytics/devguides/reporting/core/dimsmets) but doesn't let you do arbitrary aggregations like `sum of whatever`, so I'm thinking we'll just want to disable all of our normal aggregations (count, sum, etc) and populate the menu with the built-in metrics.
- [x] Allow drivers to specify the `display-name` and `description` for tables and columns. Currently they show up as `Ga-14day Users` etc.
- ~~Allow drivers to specify "built-in" segments and metrics (currently they are hard-coded)~~
- [x] What to do instead of defaulting to "Raw Data", since GA doesn't support raw data queries?
- [x] Figure out how GA accounts, properties and views map to Metabase databases, schemas, and tables
  - decided that a Metabase DB should correspond to a GA account, and every property + view combination should be a table. Not using schemas.
  - e.x. property "Metabase" property with views "All Web Site Data" and "No Bots" should create "tables" called "Metabase" and "Metabase (No Bots)"
- Query processor, translate MB query language to GA requests:
  - [x] `start-date` and `end-date`
  - [x] `metrics` (currently only supports a single metric due to #3364 but could easily support multiple)
  - [x] `dimensions`
    - [x] basic dimensions
    - [x] smart `Date` dimension that uses our date groupings
    - [x] hide redundant date fields?
    - [x] what to do about granularities that GA doesn't support? (`minute`, `day-of-year`, `quarter`, `quarter-of-year`)
  - [x] `sort`
    - [x] dimension sort
    - [x] metric sort
  - [x] `filters`
    - [x] dimension filters
    - [x] metric filter
      - ~~do we want to expose metric filtering?~~ Not yet.
    - [x] date filters using our date widget
    - [x] hide redundant date fields?
  - [x] `segment`
  - [x] `max-results`
  - ~~other fields like `start-index`, `samplingLevel`, `include-empty-rows`?~~
- [x] "native" queries
- [x] testing 😬
- [x] some relative date filters aren't being translated correctly
- [ ] we should validate that the "view"/"profile" actually belongs to the db's "account"
- [ ] docs w/ mention of rate limits (https://developers.google.com/analytics/devguides/reporting/mcf/v3/limits-quotas)

### CODE REVIEW

**Comment by mazameli** [comment]
I'm excited for this!

**Comment by salsakran** [comment]
Feels weird to have a toggle for `Enable in-depth database analysis` on this.

**Comment by wyvernnot** [comment]
This is great, given that we currently use a script to import GA metrics into MondogDB daily and then visualize it in Metabase.

Here is the script:

https://metabase-scripts.github.io/ga-scripts/import.html

**Comment by tlrobinson** [comment]
@camsaul It looks like "native" queries broke at some point, and I'm not sure why. I'm getting `Input to normalize-token does not match schema: [(named (named (not (matches-some-precondition? nil)) "Keyword or string") token)]`

**Comment by camsaul** [comment]
@tlrobinson weird. I'll look into it 👍

**Comment by tlrobinson** [comment]
@wyvernnot It would be great if you could try this branch and let us know what you think!

**Comment by camsaul** [comment]
@tlrobinson The linter is failing with some sort of `Required module not found` error

```
$ flow check 
frontend/src/metabase/services.js:6
  6: import getGAMetadata from "promise?global!metabase/lib/ga-metadata";
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ promise?global!metabase/lib/ga-metadata. Required module not found
```

File: src/metabase/driver/googleanalytics/query_processor.clj [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+
+(defn- handle-order-by [{:keys [order-by] :as query}]
+  (when order-by
+    {:sort (s/join "," (for [{:keys [field direction]} order-by]
+                         (str (case direction
+                                :ascending  ""
+                                :descending "-")
+                              (cond
+                                (instance? DateTimeField field) (unit->ga-dimension (:unit field))
+                                (instance? AgFieldRef field)    (get-in query [:aggregation :metric-name])
```
Comment (camsaul) [inline src/metabase/driver/googleanalytics/query_processor.clj]: @tlrobinson what is it you're trying to do here?

```clojure
(get-in query [:aggregation :metric-name])
```

**Comment by camsaul** [comment]
🎉

**Pull request merged**